### PR TITLE
Consensus: Sync speedup

### PR DESCRIPTION
**Pull request by kaar2**
I'm reopening the PR from the old repo. 
I've added tests and changed the default values of `MAX_DOWNLOAD_TIME` (from 10 seconds to 2 seconds) and `MAX_UNFINISHED_JOBS` (from 16 to 256.) Those values yielded best results in my tests.

### CODE REVIEW

**Comment by orogvany** [comment]
Do you think these tunings are appropriate for all hardware/bandwidth?

Anyway, approved via last review

**Comment by kaar2** [comment]
> Do you think these tunings are appropriate for all hardware/bandwidth?

I guess I can add those parameters to the config file, so everyone can tune it for his hardware and bandwidth but I think that for 99% of the users these will work just fine. Anyway, even in the case of a really bad bandwidth, this may result in a slower sync wbut the sync will still work (a timed-out block is added back to the download queue, but unless there is an error, the block will arrive eventually. So in the worst case there will be several requests of the same block). 

For very large blocks 2 seconds isn't enough, but the vast majority of the blocks contain only a few transactions. I would say that when we get to the point where the average block size is > 2MB we might need to change this.

**Comment by kaar2** [comment]
Well the fast sync needs atleast 200 blocks as it first validates all the hashes in the validator set and only then adds the blocks to the chain. About the block timeout, I thought the same at first but I noticed a dramatic boost in performance when I changed it from 10 to 2 seconds. I guess 5 seconds might work as well.

**Comment by orogvany** [comment]
Ok.  I trust your tests.  I just worry that this value will need tuning as block size increases over time.  Maybe a configuration value (with default if not set) would be worthwhile.  I don't think that should block this PR though.

File: src/test/java/org/semux/consensus/SemuxSyncTest.java [hunk 1]
```diff
@@ -206,4 +217,248 @@ public void testCheckpoints() {
...
+        SemuxSync sync = spy(new SemuxSync(kernelRule.getKernel()));
+        MessageQueue msgQueue = mock(MessageQueue.class);
+        Channel channel = mock(Channel.class);
+        when(channel.getRemoteAddress()).thenReturn(new InetSocketAddress("0.0.0.0", 5161));
+        when(channel.getMessageQueue()).thenReturn(msgQueue);
+
+        TreeSet<Pair<Block, Channel>> toProcess = new TreeSet<>(
+                Comparator.comparingLong(o -> o.getKey().getNumber()));
+
+        for (int i = 0; i < 200; i++) {
```
Comment (orogvany) [inline src/test/java/org/semux/consensus/SemuxSyncTest.java]: This and the 199 below should probably at least be static constants (or config values later on).

'Magic numbers' are an anti-pattern.

There's nothing special about 200 afaik, just a convenient number for sync.

File: src/test/java/org/semux/consensus/SemuxSyncTest.java [hunk 2]
```diff
@@ -206,4 +217,248 @@ public void testCheckpoints() {
...
+            Block block = null;
+            ;
```
Comment (orogvany) [inline src/test/java/org/semux/consensus/SemuxSyncTest.java]: errant semicolon.

File: src/test/java/org/semux/consensus/SemuxSyncTest.java (hunk 1 above)
Comment (kaar2) [reply src/test/java/org/semux/consensus/SemuxSyncTest.java]: Yes thank you. I was playing with the numbers and forgot to change it back to getValidatorUpdateInterval().

**Comment by kaar2** [comment]
@orogvany I've added all sync parameters to the config so they can be easily tuned in the future.

File: src/main/java/org/semux/consensus/SemuxSync.java [hunk 3]
```diff
@@ -315,16 +336,61 @@ private void process() {
...
+            if (!fastSync) {
+                // fastSync value is updated at the beginning of each set
+                if (latest % config.getValidatorUpdateInterval() == 0) {
+                    if (target.get() >= latest + config.getValidatorUpdateInterval()) {
+                        toFinalize.clear();
+                        currentSet.clear();
+                        lastBlockInSet = latest + config.getValidatorUpdateInterval();
+                        fastSync = true;
+                    }
+                }
```
Comment (semuxgo) [inline src/main/java/org/semux/consensus/SemuxSync.java]: I'm concerned that if more and more people are using `fastsync`, votes for blocks inside an interval are less-likely validated. Eventually, we may never be able to sync the blocks in `[start_block, last_block_of_interval]`

**Comment by orogvany** [comment]
Why would we not be able to sync those other blocks?  Clients would still hold them, just trusting they were valid, ya?

**Comment by kaar2** [comment]
I don't see how this could be abused either. `fastSync` is reset every validator interval and gets activated only if `target` >= height of the last block in the interval. Blocks on an ongoing validator round are synced in the old fashion and are fully validated.

If `fastSync` is activated, it first fully validates the last block in the interval before applying any blocks to the chain. Once it's validated, it also validates all previous blocks as their hashes can be compared against their children `parentHash` field. The only weakness (security wise) I can see compared to the normal sync is that if someone manages to create a fake block with a hash identical to the last block's `parentHash`, he might be able to fake all the blocks in the interval, instead of just one block. However this scenario is very unlikely and if someone could do that there are many other points of failure that can be exploited.

@semuxgo, Could you please elaborate on how exactly can we get to a point where we are unable to sync?

**Comment by semuxgo** [comment]
@karr2 the votes are not hashed.

**Comment by kaar2** [comment]
@semuxgo yes that is correct. But why do those votes matter? Are you concerned that we may need those votes in a future implementation and they might be lost if everyone start using fastSync?

**Comment by orogvany** [comment]
This has stagnated a bit.  I vote we approve with default=off so we at least have it available for folks initial sync

**Comment by semuxgo** [comment]
LGTM. Finished syncing in 40 minutes.

**Comment by semuxgo** [comment]
Merged

**Pull request closed without merging**